src/CoreToForSyDeIR.hs

```diff
@@ -229,54 +257,82 @@ genSignalName context =
       newContext = context {nameCounter = counter + 1}
    in (newName, newContext)
 
+-- | Helper function for `createSignals` which gets the output rate for the
+-- source of a signal. If the id is associated with a global input or delay
+-- then the returned output rate is 1.
 getSourceRate :: TranslationContext -> String -> Int
 getSourceRate context id =
   if elem id (systemInputs context)
     then 1
-    else getRateFromConstructors id (map snd (constructors context))
-
-getRateFromConstructors :: String -> [IRConstructor] -> Int
-getRateFromConstructors id list = case list of
-  [] -> -1
-  prHead : prTail -> case prHead of
-    IRDelay prId _ (_, _) ->
-      if prId == id
-        then 1
-        else getRateFromConstructors id prTail
-    IRActor prId _ _ (_, _) ->
-      if prId == id
-        then error ("TODO: implement getRateFromConstructors for actors " ++ prId)
-        else getRateFromConstructors id prTail
+    else aux (map snd (constructors context))
+  where
+    aux :: [IRConstructor] -> Int
+    aux list = case list of
+      [] -> error ("getSourceRate: " ++ id ++ " not in a valid constructor")
+      pcHead : pcTail -> case pcHead of
+        IRDelay pcId _ (_, _) ->
+          if pcId == id
+            then 1
+            else aux pcTail
+        IRActor pcId _ _ (_, _) ->
+          -- If the id is associated with a actor a value is only returned if
+          -- the identified actor has one output, otherwise the function
+          -- returns an error.
+          -- NOTE: This might be fine since if an actor has multiple outputs it
+          -- would have been represented by binder with an optional int. Will
+          -- require confirmation by exploring complex net lists.
+          if pcId == id
+            then
+              let outputRates = case (lookup pcId (pcRates context)) of
+                    Just (_, rates) -> rates
+                    Nothing -> error ("getSourceRate: No rates found for actor " ++ pcId)
+               in case outputRates of
+                    [] -> error ("getSourceRate: Empty output rates for actor " ++ pcId)
+                    [i] -> i
+                    _ -> error ("getSourceRate: More than one output rate for actor " ++ pcId)
+            else aux pcTail
 
+-- | Creates actors and delays by pattern matching based on the number of
+-- inputs to the top function, represented by `Lam`, and inputs to the process
+-- constructor, represented by `App`. If it cannot match an actor or delay then
+-- it creates a function.
 translateCoreExpr :: TranslationContext -> CoreBndr -> CoreExpr -> TranslationContext
 translateCoreExpr context binder expr = case expr of
```

Comment:
Eventually (probably not now) it might be nice to automatically construct the pattern matching from just the number of inputs and outputs, so we could provide a list, like [..., ("actor21SDF", 2, 1), ...] and it would match all of them for us.

Reply:
Related to this, it might be nice to have a comment on why we have the number of Lams and Apps. It would be nice to have a way of adding new ones without having to decipher the Core each time.